crates/ruma-serde/tests/empty_strings.rs

```diff
@@ -25,9 +36,29 @@ fn empty_se() {
 }
 
 #[test]
-fn empty_de() {
+fn empty_de_string_none() {
     let string = StringStruct { x: None };
     let none = NoneStruct { x: None };
     assert_eq!(from_json_value::<StringStruct>(json!({"x": ""})).unwrap(), string);
+    assert_eq!(from_json_value::<StringStruct>(json!({})).unwrap(), string);
     assert_eq!(from_json_value::<NoneStruct>(json!({})).unwrap(), none);
 }
+
+#[test]
+fn empty_de_string_some() {
+    let string = StringStruct { x: Some("foo".into()) };
+    assert_eq!(from_json_value::<StringStruct>(json!({"x": "foo"})).unwrap(), string);
+}
+
+#[test]
+fn empty_de_int_none() {
+    let expected = IntStruct { x: None };
+    assert_eq!(from_json_value::<IntStruct>(json!({"x": ""})).unwrap(), expected);
+    assert_eq!(from_json_value::<IntStruct>(json!({})).unwrap(), expected);
+}
+
+#[test]
+fn empty_de_int_some() {
+    let expected = IntStruct { x: Some(1) };
+    assert_eq!(from_json_value::<IntStruct>(json!({"x": 1})).unwrap(), expected);
```

Comment:
I think this cannot be made to work with how `empty_string_as_none` works. We could consider switching to `FromStr` instead of `string.into_deserializer()` to make this work, but for the use cases inside Ruma, [`StringDeserializer`](https://docs.rs/serde/1.0.130/serde/de/value/struct.StringDeserializer.html) works just fine.

Maybe you could update the test to use `ruma_identifiers::UserId` instead of `i32`? That would reflect its usage in this codebase, and later on we could evaluate whether changing the implementation makes sense.

As for the axum case that lead to this PR, I suggest you send a PR updating the example from `T: Deserialize<'de>` to `T: FromStr` and [from `T::deserialize`](https://github.com/tokio-rs/axum/blob/394d4b14f6fa03a3ec80039d9d0e736eae6a514b/examples/query-params-with-empty-strings/src/main.rs#L49) to `T::from_str`.

Reply:
I've updated the test to use `UserId`, now it passes. Will look into the suggested implementation changes tomorrow :zzz: